src/libraries/System.Linq/src/System/Linq/Select.SpeedOpt.cs

```diff
@@ -71,11 +65,73 @@ public int GetCount(bool onlyIfCheap)
 
                 return count;
             }
+
+            public override TResult? TryGetElementAt(int index, out bool found)
+            {
+                if (index >= 0)
+                {
+                    IEnumerator<TSource> e = _source.GetEnumerator();
+                    try
+                    {
+                        while (e.MoveNext())
+                        {
+                            if (index == 0)
+                            {
+                                found = true;
+                                return _selector(e.Current);
+                            }
+
+                            index--;
+                        }
+                    }
+                    finally
+                    {
+                        (e as IDisposable)?.Dispose();
+                    }
+                }
+
+                found = false;
+                return default;
+            }
+
+            public override TResult? TryGetFirst(out bool found)
+            {
+                using IEnumerator<TSource> e = _source.GetEnumerator();
+                if (e.MoveNext())
+                {
+                    found = true;
+                    return _selector(e.Current);
+                }
+
+                found = false;
+                return default;
+            }
+
+            public override TResult? TryGetLast(out bool found)
+            {
+                using IEnumerator<TSource> e = _source.GetEnumerator();
+
+                if (e.MoveNext())
+                {
+                    found = true;
+                    TSource last = e.Current;
+
+                    while (e.MoveNext())
+                    {
+                        last = e.Current;
+                    }
+
+                    return _selector(last);
+                }
+
+                found = false;
+                return default;
+            }
         }
 
-        private sealed partial class SelectArrayIterator<TSource, TResult> : IPartition<TResult>
+        private sealed partial class ArraySelectIterator<TSource, TResult>
```

Comment:
Similar to the above, this and other iterators _could_ be made to implement IList.

Reply:
I'd done that in https://github.com/dotnet/runtime/pull/88249, and in https://github.com/dotnet/runtime/pull/88249#discussion_r1248073865 you wrote "I suspect this can lead to very difficult to diagnose regressions and it's probably not worth the perf improvements." Change of heart?